### PR TITLE
Schemas validation and easy creation

### PR DESCRIPTION
**Pull request by udgover**
> [!IMPORTANT] 
This PR removes `add_text` method in Observable class. It breaks private plugins relying on this method to create observables.

# New features

#### Overview

Validators are now enforced for observables listed below. Validators ensure that provided value is relevant and that an for example an `IPv4` observable cannot be built from a `Url`.

* bic
* email
* hostname
* iban
* ipv4
* ipv6
* mac_address
* md5
* path
* sha1
* sha256
* url

> [!NOTE]
Validators enforcement also impacts `url` observable creation which requires a scheme to be valid. This means that `example.com/foobar` is no longer a valid `url`.  

#### Implementation details

Value validation relies on pydandic `field_validator` decorator. Any observable implementation that requires validation must implement `validate_value` class method with `@field_validator("value")` decorator. `validate_value` is also where refang of the value takes place. If the value is not valid, ValueError exception must be raised. Otherwise, the value (modified or not) must be returned as a string.

```python
class IPv4(observable.Observable):
    type: Literal["ipv4"] = "ipv4"

    @field_validator("value")
    @classmethod
    def validate_value(cls, value: str) -> str:
        value = observable.refang(value)
        if not validators.ipv4(value):
            raise ValueError("Invalid ipv4 address")
        return value
```

### Helper functions

To ease observables creation, new functions have been implemented in `observable` module:

```python
>>> from core.schemas import observable
>>> observable.guess_type("192.168.1[.]2")
'ipv4'
>>> observable.create(value="192.168.1[.]2")
IPv4(value='192.168.1.2', created=datetime.datetime(2024, 10, 29, 23, 57, 4, 504655, tzinfo=datetime.timezone.utc), context=[], last_analysis={}, type='ipv4', id=None, tags={}, root_type='observable')
>>> observable.find(value="192.168.1[.]2")
>>> observable.save(value="192.168.1[.]2")
IPv4(value='192.168.1.2', created=datetime.datetime(2024, 10, 29, 23, 57, 20, 82795, tzinfo=TzInfo(UTC)), context=[], last_analysis={}, type='ipv4', id='5638348', tags={}, root_type='observable')
>>> observable.find(value="192.168.1[.]2")
IPv4(value='192.168.1.2', created=datetime.datetime(2024, 10, 29, 23, 57, 20, 82795, tzinfo=TzInfo(UTC)), context=[], last_analysis={}, type='ipv4', id='5638348', tags={}, root_type='observable')
```

#### Guess observable type from value

```python
guess_type(value: str)
``` 

Guess which observable type can validate `value` string. If the value can be validated by an observable, its type is returned as a string. If the value can't be guessed, `None` is returned.

#### Create an observable from string

Create an observable from the given string value without saving it in the database. It's up to the caller to then call `save()` on the returned object to save it in the database.

```python
create(value: str, type: str | None = None, **kwargs)
```

* Params
   * `value`: a string defining the observable.
   * `type`: an optional string defining the type of the observable. If `type` is not provided, it will be guessed by calling `guess_type`.
   * `kwargs`: a dictionary defining further attributes depending on the type of the observable.

* Returns
If the function call succeeds, the created observable is returned. Otherwise, an `ValueError` exception is raised.

#### Save an observable from string

Use `create` function and save observable in the database. Tag the resulting observable with the list of tags if defined. 

```python
save(*, value: str, type: str | None = None, tags: List[str] = None, **kwargs) -> ObservableTypes
```

* Params
   * `value`: a string defining the observable.
   * `type`: an optional string defining the type of the observable. If `type` is not provided, it will be guessed by calling `guess_type`.
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observable.
   * `kwargs`: a dictionary defining further attributes depending on the type of the observable.

* Returns
If the function call succeeds, the saved observable is returned. Otherwise, an `ValueError` exception is raised.

#### Create observables from text

Create observables from the given text. Each line must contain one observable. Guess the type of the observable and create it. Created observable are not saved in the database.

```python
create_from_text(text: str)
```

* Params
   * `text`: a string containing one observable per line.
* Returns
A tuple containing a list of created observables and a list of string for unguessable lines.

#### Save observables from text

Use `create_from_text` function and save observables in the database. Tag the resulting observables with the list of tags if defined. 

```python
save_from_text(text: str, tags: List[str] = None)
```

* Params
   * `text`: a string containing one observable per line.
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observables.
* Returns
A tuple containing a list of saved observables and a list of string for unguessable lines.

#### Create observables from file

Create observables from the given file. Each line must contain one observable. Guess the type of the observable and creates it. Created observables are not saved in the database.

```python
create_from_file(file: FileLikeObject)
```

* Params
   * `file`: Represents a file in different ways:
      * str, bytes or os.PathLike: the provided file path will be opened for reading lines
      * io.IOBase: any file object either created from `open` or based on `io` classes
      * tempfile.SpooledTemporaryFile: defined to support API endpoint from UploadFile FastAPI.
* Returns
A tuple containing a list of created observables and a list of string for unguessable lines.

#### Save observables from file

Use `create_from_file` function and save observables into the database. Tag the resulting observables with the list of tags if defined. 

```python
save_from_file(file: FileLikeObject, tags: List[str] = None)
```

* Params
   * `file`: Represents a file in different ways:
      * str, bytes or os.PathLike: the provided file path will be opened for reading lines
      * io.IOBase: any file object either created from `open` or based on `io` classes
      * tempfile.SpooledTemporaryFile: defined to support API endpoint from UploadFile FastAPI.
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observables.
* Returns
A tuple containing a list of saved observables and a list of string for unguessable lines.

#### Create observables from a url

Create observables from the given url. Each line must contain one observable. Guess the type of the observable and create it. Created observables are not saved in the database.

```python
create_from_url(url: str)
```

* Params
   * `url`: a string defining the URL to fetch the content from.
* Returns
A tuple containing a list of created observables and a list of string for unguessable lines.

#### Save observables from a url

Use `create_from_url` function and save observables in the database. Tag the resulting observabls with the list of tags if defined. 

```python
save_from_url(url: str, tags: List[str] = None)
```

* Params
   * `url`: a string defining the URL to fetch the content from.
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observables.
* Returns
A tuple containing a list of saved observables and a list of string for unguessable lines.

#### Find an observable object

```python
find(value: str, **kwargs)
```

Find an observable in the database matching the string `value` and optional fields represented by `kwargs`. This function automatically refangs defanged `value` before querying the database. Return an observable object.

### Create an entity

Create an entity from the given name and type without saving it to the database.

```python
create(name: str, type: str, **kwargs)
```

* Params
   * `name`: a string defining the name of the entity
   * `type`: a string defining the type of the entity
   * `kwargs`: a dictionary defining further attributes depending on the type of the entity.
* Returns
If the function call succeeds, the created entity is returned. Otherwise, an `ValueError` exception is raised.

### Save an entity

```python
save(name: str, type: str, tags: List[str] = None, **kwargs)
```

Use `create` function and save entity in the database. Tag the resulting entity with the list of tags if defined. 

* Params
   * `name`: a string defining the name of the entity to create
   * `type`: a string defining the type of the entity to create
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observables.
   * `kwargs`: a dictionary defining further attributes depending on the type of the entity.
* Returns
If the function call succeeds, the saved entity is returned. Otherwise, an `ValueError` exception is raised.

### Create an indicator

Create an indicator from the given name, type, pattern and diamond without saving it to the database.

```python
create(name: str, type: str, pattern: str, diamond: DiamondModel, **kwargs)
```

* Params
   * `name`: a string defining the name of the indicator
   * `type`: a string defining the type of the indicator
   * `pattern`: a string defining the pattern of the indicator
   * `diamond` a string defining the diamond model of the indicator
   * `kwargs`: a dictionary defining further attributes depending on the type of the indicator.
* Returns
If the function call succeeds, the created indicator is returned. Otherwise, an `ValueError` exception is raised.

### Save an entity

```python
save(name: str, type: str, pattern: str, diamond: DiamondModel, tags: List[str] = None, **kwargs)
```

Use `create` function and save indicator in the database. Tag the resulting indicator with the list of tags if defined. 

* Params
   * `name`: a string defining the name of the indicator
   * `type`: a string defining the type of the indicator
   * `pattern`: a string defining the pattern of the indicator
   * `diamond` a string defining the diamond model of the indicator
   * `tags`: an optional list of strings corresponding to the tags to add to the saved observables.
   * `kwargs`: a dictionary defining further attributes depending on the type of the entity.
* Returns
If the function call succeeds, the saved indicator is returned. Otherwise, an `ValueError` exception is raised.

## API

New observables endpoints have been implemented to support the new save and create functions:

### Import observables  from text

Save observables from a text containing one observable per line. Tag with defined tags.

* Endpoint `POST api/v2/observables/import/text`
* Usage:
```python
import requests

observables = """1.1.1[.]1
8.8.8.8
tomchop[.]me
google.com
http://google.com/
http://tomchop[.]me/
d41d8cd98f00b204e9800998ecf8427e
da39a3ee5e6b4b0d3255bfef95601890afd80709
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"""

requests.post("api/v2/observables/import/text", json={"text": observables, "tags": ["tag1", "tag2"]})
```

### Import observables from file

Save observables from a file containing one observable per line. Tag with defined tags.

* Endpoint `POST api/v2/observables/import/file`
* Usage:
```python
import requests

with open(path_to_file, "rb") as file:
   requests.post("/api/v2/observables/import/file", files={"file": file}, data={"tags": ["tag1", "tag2"]})
```

### Import observables from URL

Save observables from a text containing one observable per line. Tag with defined tags.

* Endpoint `POST api/v2/observables/import/file`
* Usage:
```python
import requests

requests.post("api/v2/observables/import/text",  json={"url": "https://example.com", "tags": ["tag1", "tag2"]})
```

### CODE REVIEW

File: core/schemas/entity.py [hunk 1]
```diff
@@ -78,3 +78,31 @@ def add_context(
...
+    indicator_obj = create(type, **kwargs).save()
+    if tags:
+        indicator_obj.tag(tags)
+    return indicator_obj
+
+
+def get(**kwargs) -> "EntityTypes":
+    if "name" not in kwargs:
+        raise ValueError("value is a required field for an entity")
+    return Entity.find(**kwargs)
```
Comment (tomchop) [inline core/schemas/entity.py:108]: `get` is another database_arango function, I suggest renaming it to `find` to make it more transparent.

```suggestion
def find(**kwargs) -> "EntityTypes":
    if "name" not in kwargs:
        raise ValueError("name is a required field for an entity")
    return Entity.find(**kwargs)
```

File: core/schemas/indicator.py [hunk 2]
```diff
@@ -88,3 +88,31 @@ def search(cls, observables: list[str]) -> list[tuple[str, "Indicator"]]:
...
+    indicator_obj = create(type, **kwargs).save()
+    if tags:
+        indicator_obj.tag(tags)
+    return indicator_obj
+
+
+def get(**kwargs) -> "IndicatorTypes":
+    if "name" not in kwargs:
+        raise ValueError("value is a required field for an indicator")
+    return Indicator.find(**kwargs)
```
Comment (tomchop) [inline core/schemas/indicator.py:118]:
```suggestion
def find(**kwargs) -> "IndicatorTypes":
    if "name" not in kwargs:
        raise ValueError("name is a required field for an indicator")
    return Indicator.find(**kwargs)
```

File: core/schemas/observable.py [hunk 3]
```diff
@@ -128,8 +97,63 @@ def delete_context(
...
+    Create an observable object without saving it to the database.
+
+    kwargs must contain a "value" field representing the value of the observable.
+
+    if kwargs does not contain a "type" field, type will be automatically
+    determined based on the value. If the type is not recognized, an observable
+    of type generic will be created.
+    """
+    if "value" not in kwargs:
+        raise ValueError("name is a required field for an observable")
```
Comment (tomchop) [inline core/schemas/observable.py:128]:
```suggestion
    if "value" not in kwargs:
        raise ValueError("value is a required field for an observable")
```

File: core/schemas/observable.py [hunk 4]
```diff
@@ -128,8 +97,63 @@ def delete_context(
...
+
+def create(type: str | None = None, **kwargs) -> ObservableTypes:
+    """
+    Create an observable object without saving it to the database.
+
+    kwargs must contain a "value" field representing the value of the observable.
+
+    if kwargs does not contain a "type" field, type will be automatically
+    determined based on the value. If the type is not recognized, an observable
+    of type generic will be created.
```
Comment (tomchop) [inline core/schemas/observable.py:125]:
```suggestion
    Create an observable object without saving it to the database.

    kwargs must contain a "value" key representing the value of the observable.

    if kwargs does not contain a "type" key, type will be automatically
    determined based on the value. If the type is not recognized, an observable
    of type generic will be created.
```

File: core/schemas/observable.py [hunk 5]
```diff
@@ -128,8 +97,63 @@ def delete_context(
...
+
+    tags is an optional list of tags to add to the observable.
+    """
+    observable_obj = create(**kwargs).save()
+    if tags:
+        observable_obj.tag(tags)
+    return observable_obj
+
+
+def get(**kwargs) -> ObservableTypes:
```
Comment (tomchop) [inline core/schemas/observable.py:156]:
```suggestion
def find(**kwargs) -> ObservableTypes:
```

File: core/web/apiv2/observables.py [hunk 6]
```diff
@@ -191,23 +192,15 @@ async def bulk_add(request: NewBulkObservableAddRequest) -> BulkObservableAddRes
...
-                    observable = observable.tag(new_observable.tags)
-            except (ValueError, RuntimeError):
-                response.failed.append(new_observable.value)
-                continue
+        try:
+            observable = _observable.save(
+                type=new_observable.type,
+                value=new_observable.value,
+                tags=new_observable.tags,
+            )
```
Comment (tomchop) [inline core/web/apiv2/observables.py:200]: I'd rather rename the variable than use an import alias, WDYT?

File: plugins/analytics/public/censys.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 from core import taskmanager
 from core.config.config import yeti_config
 from core.schemas import indicator, task
-from core.schemas.observable import Observable
+from core.schemas import observable as observable
```
Comment (tomchop) [inline plugins/analytics/public/censys.py:9]: Probably don't need the `as observable` here ?

File: plugins/analytics/public/passive_total.py [hunk 8]
```diff
@@ -6,6 +6,7 @@
 
 from core import taskmanager
 from core.config.config import yeti_config
+from core.schemas import observable as _observable
```
Comment (tomchop) [inline plugins/analytics/public/passive_total.py:9]: Same comment as before

File: tests/analytics_test.py [hunk 9]
```diff
@@ -7,7 +7,8 @@
 from parameterized import parameterized
 
 from core import database_arango
-from core.config.config import yeti_config
+
+# from core.config.config import yeti_config
 from core.schemas import indicator, observable
```
Comment (tomchop) [inline tests/analytics_test.py:12]:
```suggestion

from core.schemas import indicator, observable
```

**Review by tomchop — CHANGES_REQUESTED** [review]

File: core/schemas/observable.py [hunk 10]
```diff
@@ -128,8 +102,179 @@ def delete_context(
...
+
+def create(*, value: str, type: str | None = None, **kwargs) -> ObservableTypes:
+    """
+    Create an observable object without saving it to the database.
+
+    value argument representing the value of the observable.
+
+    if kwargs does not contain a "type" field, type will be automatically
+    determined based on the value. If the type is not recognized, an observable
+    of type generic will be created.
```
Comment (tomchop) [inline core/schemas/observable.py:130]: Please adjust docstring here

File: core/schemas/observable.py [hunk 11]
```diff
@@ -128,8 +102,179 @@ def delete_context(
...
+        f = file
+    else:
+        raise ValueError("Invalid file type")
+    observables = list()
+    unknown = list()
+    for line in f.readlines():
+        if isinstance(line, bytes):
+            line = line.decode("utf-8").strip()
+        else:
+            line = line.strip()
```
Comment (tomchop) [inline core/schemas/observable.py:222]:
```suggestion
        if isinstance(line, bytes):
            line = line.decode("utf-8")
        line = line.strip()
```

File: core/web/apiv2/observables.py [hunk 12]
```diff
@@ -282,36 +280,83 @@ async def add_text(request: AddTextRequest) -> ObservableTypes:
...
     its type."""
     try:
-        return Observable.add_text(request.text, request.tags)
+        return observable.save(value=request.text, tags=request.tags)
     except ValueError as error:
         raise HTTPException(status_code=400, detail=str(error))
 
 
+@router.post("/import/text")
+async def import_from_text(request: ImportTextRequest) -> BulkObservableAddResponse:
```
Comment (tomchop) [inline core/web/apiv2/observables.py:289]: Can we just deprecate `add_text` in this case? They seem to be covering pretty much exaclty the same use-case?

If we're worried about breaking APIs, then I'm fine keeping add_text, but maybe we can just add the `@router.post` handler to this function instead?

(also, does FastAPI handle deprecation notices in endpoints?)

File: tests/analytics_test.py [hunk 13]
```diff
@@ -97,6 +96,10 @@ def test_shodan_query_with_various_limits(self, limit, expected_count, mock_shod
         mock_shodan.return_value = mock_shodan_api
 
         os.environ["YETI_SHODAN_API_KEY"] = "test_api_key"
+        if limit:
+            os.environ["YETI_SHODAN_RESULT_LIMIT"] = str(limit)
+        else:
+            os.environ["YETI_SHODAN_RESULT_LIMIT"] = ""
```
Comment (tomchop) [inline tests/analytics_test.py:102]: Do we want this or to `del` the env variable?

**Review by tomchop — CHANGES_REQUESTED** [review]
Some minor comments, but otherwise this looks great!

File: core/schemas/observable.py (hunk 10 above)
Comment (udgover) [reply core/schemas/observable.py]: fixed in 900075a392262e00e39fc80decbae7eda306ac94

File: core/schemas/observable.py (hunk 11 above)
Comment (udgover) [reply core/schemas/observable.py]: fixed in ef2578ac6e642a193dd0c4f67f39c69dae4766eb

File: core/web/apiv2/observables.py (hunk 12 above)
Comment (udgover) [reply core/web/apiv2/observables.py:289]: fixed in 8417c1187e3db0f0c5d7d86d2eeb1bba47ec6b7d

File: tests/analytics_test.py (hunk 13 above)
Comment (udgover) [reply tests/analytics_test.py]: fixed in 48a6ce0820487573af5c33c7185482433c533901

**Review by udgover — COMMENTED** [review]
Fixed your comments

File: core/schemas/observable.py [hunk 14]
```diff
@@ -24,6 +28,7 @@ class ObservableType(str, Enum): ...
 
 ObservableTypes = ()
 TYPE_MAPPING = {}
+FileLikeObject = Union[str, os.PathLike, IO, tempfile.SpooledTemporaryFile]
```
Comment (tomchop) [inline core/schemas/observable.py:31]: Let's settle on one notation (at least throughout the file); either use `Union` or use ` | ` (I have a slight preference for the latter)

File: plugins/feeds/public/lolbas.py [hunk 15]
```diff
@@ -126,3 +163,5 @@ def format_commands(self, commands: list[dict[str, str]]) -> str:
 
 
 taskmanager.TaskManager.register_task(LoLBAS)
+taskmanager.TaskManager.register_task(LoLBAS)
+taskmanager.TaskManager.register_task(LoLBAS)
```
Comment (tomchop) [inline plugins/feeds/public/lolbas.py:167]: Duplicate registration?

**Review by tomchop — APPROVED** [review]
Minor nits, otherwise looks good!

File: core/schemas/observable.py (hunk 14 above)
Comment (udgover) [reply core/schemas/observable.py]: Fixed in 1c5cc8f0abd23d6e1dab6589460acad05faa5452

File: plugins/feeds/public/lolbas.py (hunk 15 above)
Comment (udgover) [reply plugins/feeds/public/lolbas.py]: Fixed in 1c5cc8f0abd23d6e1dab6589460acad05faa5452

**Pull request merged**